src/loader/bgw_message_queue.c

```diff
@@ -0,0 +1,373 @@
+ * the extension. Therefore we decided to simply take an exclusive lock whenever we were modifying
+ * anything in the shared memory segment to avoid collisions.
+ */
+
+static pid_t
+queue_get_reader(MessageQueue * queue)
+{
+	pid_t		reader;
+	volatile	MessageQueue *vq = queue;
+
+	SpinLockAcquire(&vq->mutex);
+	reader = vq->reader_pid;
+	SpinLockRelease(&vq->mutex);
+	return reader;
+}
+static void
+queue_set_reader(MessageQueue * queue)
+{
+	volatile	MessageQueue *vq = queue;
+
+	if (queue_get_reader(queue) == InvalidPid)
+	{
+		queue_init();
+		SpinLockAcquire(&vq->mutex);
+		vq->reader_pid = MyProcPid;
+		SpinLockRelease(&vq->mutex);
+	}
+	else if (queue_get_reader(queue) != MyProcPid)
+		ereport(ERROR, (errmsg("only one reader for allowed for TimescaleDB background worker message queue")));
+}
+
+static void
+queue_reset_reader(MessageQueue * queue)
+{
+	volatile	MessageQueue *vq = queue;
+
+	SpinLockAcquire(&vq->mutex);
+	vq->reader_pid = InvalidPid;
+	SpinLockRelease(&vq->mutex);
+}
+
+
+/* Add a message to the queue - we can do this if the queue is not full */
+static QueueResponseType queue_add(MessageQueue * queue, BgwMessage * message)
+{
+	QueueResponseType message_sent = QUEUE_FULL;
+
+	LWLockAcquire(queue->lock, LW_EXCLUSIVE);
+	if (queue->num_elements < BGW_MQ_MAX_MESSAGES)
+	{
+		memcpy(&queue->buffer[(queue->read_upto + queue->num_elements) % BGW_MQ_MAX_MESSAGES], message, sizeof(BgwMessage));
+		queue->num_elements++;
+		message_sent = MESSAGE_SENT;
+	}
+	LWLockRelease(queue->lock);
+
+	if (queue_get_reader(queue) != InvalidPid)
+		SetLatch(&BackendPidGetProc(queue_get_reader(queue))->procLatch);
+	else
+		message_sent = READER_DETACHED;
+	return message_sent;
+}
+
+static BgwMessage * queue_remove(MessageQueue * queue)
+{
+	BgwMessage *message = NULL;
+
+	LWLockAcquire(queue->lock, LW_EXCLUSIVE);
+	if (queue_get_reader(queue) != MyProcPid)
+		ereport(ERROR, (errmsg("cannot read if not reader for TimescaleDB background worker message queue")));
+
+	if (queue->num_elements > 0)
+	{
+		message = palloc(sizeof(BgwMessage));
+		memcpy(message, &queue->buffer[queue->read_upto], sizeof(BgwMessage));
+		queue->read_upto = (queue->read_upto + 1) % BGW_MQ_MAX_MESSAGES;
+		queue->num_elements--;
+	}
+	LWLockRelease(queue->lock);
+	return message;
+}
+
+/*Construct a message*/
+static BgwMessage * bgw_message_create(BgwMessageType message_type, Oid db_oid)
+{
+	BgwMessage *message = palloc(sizeof(BgwMessage));
+	dsm_segment *seg;
+
+	seg = dsm_create(BGW_ACK_QUEUE_SIZE, 0);
+
+	*message = (BgwMessage)
+	{
+		.message_type = message_type,
+			.sender_pid = MyProcPid,
+			.db_oid = db_oid,
+			.ack_dsm_handle = dsm_segment_handle(seg)
+	};
+
+	return message;
+}
+
+
+/* Our own version of shm_mq_wait_for_attach that waits with a timeout so that should our counterparty die before attaching,
+ * we don't end up hanging.*/
+static shm_mq_result
+ts_shm_mq_wait_for_attach(MessageQueue * queue, shm_mq_handle *ack_queue_handle)
```

Comment:
This time it is here because it means it is a wrapped version of the postgres function which we changed for our own purposes. Not sure if you think there should be a different prefix to mean that.

Reply:
keep if you want, but the name is lower-case now anyway so it is not important to have the `ts_` prefix to avoid conflicts.